### PR TITLE
Fix/avx512

### PR DESCRIPTION
**Pull request by jtlap**

### CODE REVIEW

File: include/eve/module/real/core/function/regular/simd/x86/add.hpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+namespace eve::detail
+{
+  // -----------------------------------------------------------------------------------------------
+  // 128 bits implementation
+  template<real_scalar_value T, typename N>
+  EVE_FORCEINLINE wide<T, N> add_(EVE_SUPPORTS(sse2_), wide<T, N> v0, wide<T, N> v1) noexcept
+    requires x86_abi<abi_t<T, N>>
+  {
+    return v0 += v1;
+  }
```
Comment (DenisYaroshevskiy) [inline include/eve/module/real/core/function/regular/simd/x86/add.hpp:25]: Why is this necessary? I mean the base case should still work, right?

File: include/eve/module/real/core/function/regular/simd/x86/add.hpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+  // -----------------------------------------------------------------------------------------------
+  // Masked case
+  template<conditional_expr C, real_scalar_value T, typename N>
+  EVE_FORCEINLINE
+  wide<T, N> add_(EVE_SUPPORTS(sse2_), C const &cx, wide<T, N> const &v, wide<T, N> const &w) noexcept
```
Comment (DenisYaroshevskiy) [inline include/eve/module/real/core/function/regular/simd/x86/add.hpp:31]: Supports avx512?
Or `requires !abi_t<T, N>::is_wide_logical`?

File: include/eve/module/real/core/function/regular/simd/x86/div.hpp [hunk 3]
```diff
@@ -18,4 +18,31 @@ namespace eve::detail
   {
     return v0 /= v1;
   }
```
Comment (DenisYaroshevskiy) [inline include/eve/module/real/core/function/regular/simd/x86/div.hpp:20]: Why is this needed for x86 special?

File: include/eve/module/real/core/function/regular/simd/x86/popcount.hpp [hunk 4]
```diff
@@ -127,4 +128,90 @@ namespace eve::detail
...
   }
+
+  /////////////////////////////////////////////////////////////////////////////
+  // avx512
+  template<integral_scalar_value T,  typename N>
+  EVE_FORCEINLINE auto popcount_(EVE_SUPPORTS(avx512_), wide<T, N>  v) noexcept
+      requires x86_abi<abi_t<T, N>>
+  {
+    auto constexpr bitalg = spy::supports::avx512::bitalg_;
+    auto constexpr ppctdq = spy::supports::avx512::popcntdq_;
```
Comment (DenisYaroshevskiy) [inline include/eve/module/real/core/function/regular/simd/x86/popcount.hpp:139]: but we don't have this on CI.
If we don't have something on CI, adding it is a problem.

File: include/eve/module/real/core/function/saturated/simd/x86/add.hpp [hunk 5]
```diff
@@ -47,4 +47,40 @@ namespace eve::detail
...
+      else if constexpr ( current_api >= avx2 && c == category::int16x16  ) return _mm256_mask_adds_epi16(src,m,v,w);
+      else if constexpr ( current_api >= avx2 && c == category::uint16x16 ) return _mm256_mask_adds_epu16(src,m,v,w);
+      else if constexpr ( current_api >= avx2 && c == category::int8x32   ) return _mm256_mask_adds_epi8 (src,m,v,w);
+      else if constexpr ( current_api >= avx2 && c == category::uint8x32  ) return _mm256_mask_adds_epu8 (src,m,v,w);
+      else if constexpr ( c == category::int16x8                          ) return _mm_mask_adds_epi16   (src,m,v,w);
+      else if constexpr ( c == category::uint16x8                         ) return _mm_mask_adds_epu16   (src,m,v,w);
+      else if constexpr ( c == category::int8x16                          ) return _mm_mask_adds_epi8    (src,m,v,w);
+      else if constexpr ( c == category::uint8x16                         ) return _mm_mask_adds_epu8    (src,m,v,w);
+      else                                                                  return if_else(cx,eve::add(st, v, w),src);
+    }
```
Comment (DenisYaroshevskiy) [inline include/eve/module/real/core/function/saturated/simd/x86/add.hpp:83]: Why not fwd cpu?

**Review by DenisYaroshevskiy — COMMENTED** [review]
Is everything here tested already?

File: include/eve/module/real/core/function/regular/simd/x86/add.hpp (hunk 1 above)
Comment (jtlap) [reply include/eve/module/real/core/function/regular/simd/x86/add.hpp:25]: perhaps useless I did it as it was in sub

File: include/eve/module/real/core/function/regular/simd/x86/div.hpp (hunk 3 above)
Comment (jtlap) [reply include/eve/module/real/core/function/regular/simd/x86/div.hpp:20]: don't know same answer, it was there

File: include/eve/module/real/core/function/regular/simd/x86/popcount.hpp (hunk 4 above)
Comment (jtlap) [reply include/eve/module/real/core/function/regular/simd/x86/popcount.hpp:139]: ok I just wanted to see if it worked on CI, the PR was the to see if there were problems.
On my computer I have bitalg and popcntdq and I wanted to see if it was easy to add these 
At home it works enabling the options on the compiler and also just putting -march=skylake...

File: include/eve/module/real/core/function/saturated/simd/x86/add.hpp (hunk 5 above)
Comment (jtlap) [reply include/eve/module/real/core/function/saturated/simd/x86/add.hpp:83]: ok can do

**Comment by jtlap** [comment]
In fact this is not finished yet I sent the PR just to see it worked outside of avx512 and there are still some mistakes
I am on it
But all features added are tested in the cpp files

File: include/eve/module/real/core/function/regular/simd/x86/popcount.hpp (hunk 4 above)
Comment (DenisYaroshevskiy) [reply include/eve/module/real/core/function/regular/simd/x86/popcount.hpp:139]: What I mean is.

By avx512 we mean:
`AVX512 with skylake like minimum flavour (f,vl,dq,bw)`

https://jfalcou.github.io/eve/index.html

I'd prefer if we either stick by it or compile versions where it's disabled on CI as well.

Otherwise we have a code that's never tested, which means that the users might hit it and get runtime bugs.

File: include/eve/module/real/core/function/regular/simd/x86/add.hpp (hunk 1 above)
Comment (DenisYaroshevskiy) [reply include/eve/module/real/core/function/regular/simd/x86/add.hpp:25]: Maybe remove it everywhere if you don't mind?
Just: less code - easier to maintain?

File: include/eve/module/real/core/function/regular/simd/x86/popcount.hpp (hunk 4 above)
Comment (jtlap) [reply include/eve/module/real/core/function/regular/simd/x86/popcount.hpp:139]: ok I finish testing my stuffs and will remove the bitalg etc part,
but it is not never tested as it was tested on my computer who have it

File: include/eve/module/real/core/function/saturated/simd/x86/add.hpp (hunk 5 above)
Comment (jtlap) [reply include/eve/module/real/core/function/saturated/simd/x86/add.hpp:83]: done

File: include/eve/module/real/core/function/regular/simd/x86/popcount.hpp (hunk 4 above)
Comment (DenisYaroshevskiy) [reply include/eve/module/real/core/function/regular/simd/x86/popcount.hpp:139]: I meant on ci. The code needs to be contiguous tested so that the new changes don't break old behaviour.

File: include/eve/module/real/core/function/regular/simd/x86/mul.hpp [hunk 6]
```diff
@@ -13,10 +13,33 @@
...
+      return mul_(EVE_RETARGET(cpu_),cx,v,w);
+    }
+    else
+    {
+      auto src  = alternative(cx,v,as<wide<T, N>>{});
+      auto m    = expand_mask(cx,as<wide<T, N>>{}).storage().value;
+
+            if constexpr(c == category::float32x16) return _mm512_mask_mul_ps   (src,m,v,w);
+      else  if constexpr(c == category::float64x8 ) return _mm512_mask_mul_pd   (src,m,v,w);
+#ifndef SPY_COMPILER_IS_GCC
```
Comment (jfalcou) [inline include/eve/module/real/core/function/regular/simd/x86/mul.hpp:36]: you can use `spy::compiler == spy::gcc` in a `if constexpr`

File: include/eve/module/real/core/function/regular/simd/x86/rec.hpp [hunk 7]
```diff
@@ -13,53 +13,102 @@
...
+    else  if constexpr(c == category::float32x8  )
+    {
+      if constexpr(current_api >= avx2 )            return _mm256_rcp14_epi32(v);
+      else                                          return _mm256_rcp_ps(v);
+    }
+    else  if constexpr(c == category::int64x4 )
     {
-      // The error for this approximation is no more than 1.5.e-12
-      return _mm_cvtps_pd(_mm_rcp_ps(_mm_cvtpd_ps(a0)));
+      if constexpr(current_api >= avx2 )            return _mm256_rcp14_epi64(v);
```
Comment (jfalcou) [inline include/eve/module/real/core/function/regular/simd/x86/rec.hpp:31]: This function doesn't exist.
The whole rcp is only available on real so why bother with integer ?

File: include/eve/module/real/core/function/regular/simd/x86/rec.hpp [hunk 8]
```diff
@@ -13,53 +13,102 @@
...
+      else                                          return _mm256_cvtps_pd(_mm_rcp_ps(_mm256_cvtpd_ps(v)));
+    }
+    else  if constexpr(c == category::float32x4  )
+    {
+      if constexpr(current_api >= avx2 )            return _mm_rcp14_epi32(v);
```
Comment (jfalcou) [inline include/eve/module/real/core/function/regular/simd/x86/rec.hpp:36]: Same for rcp14_epI32 and rcp14_epi64

File: include/eve/module/real/core/function/regular/generic/fma.hpp [hunk 9]
```diff
@@ -37,4 +37,16 @@ namespace eve::detail
...
+
+  //================================================================================================
+  // Masked case
+  //================================================================================================
+  template<conditional_expr C, real_value T, real_value U, real_value V>
+  EVE_FORCEINLINE auto fma_(EVE_SUPPORTS(cpu_), C const &cond, T const &a, U const &b, V const &c) noexcept
+  requires properly_convertible<U, V, T>
+  {
+    using r_t =  common_compatible_t<T, U, V>;
+    return mask_op(  cond, eve::fma, r_t(a), r_t(b), r_t(c));
```
Comment (jfalcou) [inline include/eve/module/real/core/function/regular/generic/fma.hpp:49]: Why do you do a conversion here and not in the other fma like mask functions ?

**Review by jfalcou — CHANGES_REQUESTED** [review]
I think we need to make all the if_else in mask just retarget to CPU just in case one day we change this strategy.

File: include/eve/module/real/core/function/regular/simd/x86/rec.hpp (hunk 7 above)
Comment (jtlap) [reply include/eve/module/real/core/function/regular/simd/x86/rec.hpp:31]: mistakes corrected

**Review by jfalcou — APPROVED** [review]
